server/src/main/java/org/opensearch/ingest/IngestService.java

```diff
@@ -109,6 +112,13 @@ public class IngestService implements ClusterStateApplier, ReportingService<Inge
 
     private static final Logger logger = LogManager.getLogger(IngestService.class);
 
+    public static final Setting<List<String>> PROCESSORS_ALLOWLIST_SETTING = Setting.listSetting(
+        "ingest.processors.allowlist",
```

Comment:
Should this be `ingest.pipeline.processors.allowlist`? Note we'll have the corresponding search variant as well. "pipeline" felt redundant for "ingest", because in the context of the code base "ingest" is kind of short hand for "ingest pipeline". That's not true in the case of "search"/"search pipepline".

Reply:
Thanks @andrross , three issue here (as I see it): 
 - the processors are coming from `IngestCommonPlugin` so we should reflect that by prefixing the setting with the `plugins` as per our conventions (and moving it to plugin/module as well):
    ```
    plugins.ingest.pipeline.processors.
    ```
 - the plugin should filter out the processors (IngestCommonPlugin::getProcessors) with respect to the settings configured, the `IngestService` should not be involved I believe since the processors could be contributed by other plugins (the service would never see these processors)
 
- may be  just using `*.allowed` is sufficient? No strong opinions vs `allowlist` but list sounds like a bit redundant

Reply:
@reta I don't think we should scope this to just the `ingest-common` plugin. This is intended to be an allowlist that opts in to a specific set of processors, regardless of which plugin they come from. My understand is that `IngestService` is the single place to register all processors from all plugins, hence the place that the "allowlist" filter should go.

Reply:
It looks very inconsistent to - why we do that in core? What if plugin is uninstalled? The settings looks detached from the context to me.

Reply:
Here's the specific use case: I want to enable some, but not all, of the processors defined in the neural-search plugin and none of the plugins in ingest-common.

Agree there are sharp edges here, but it is intended only for very customized installations and most users/operators will never need to interact with it.

Reply:
> Here's the specific use case: I want to enable some, but not all, of the processors defined in the neural-search plugin and none of the processors in ingest-common.

This is super straightforward (assuming we could not uninstall `ingestion-common`):

```
plugins.ingest.pipeline.processors.allowed = []
plugins.neural-search.pipeline.processors.allowed = [ 'some' ]
```

Reply:
> Where do you see these settings being implemented? 

In a plugin / module (as we do now, fe for `security`) 

> Do we have to duplicate this setting definition and allowlist filtering in each plugin/module?

This is not a duplicate per se - the setting is plugin specific. Otherwise it is impossible to perform any meaningful validation, the processors list could contain anything, no matter if such processor exists or not. 

Here is the reasoning:
 - user develops plugin (with ingest processors)
 - user install this plugin, things work right away
 - or, user enables some processor(s) using the plugin setting

I understand you point as well, but the core is taking too much responsibility in my opinion in this case - we essentially do not trust our users. The logical question after that - let's restrict analyzers, normalizers, ... everything essentially cuz, well, this is a risk, right?

Reply:
> core is taking too much responsibility

The core allows for registering processors by name. This setting simply ignores any processor that does not match a pre-defined list provided by the operator.

> Otherwise it is impossible to perform any meaningful validation, the processors list could contain anything

The validation as implemented in this PR is that if a processor is listed as "allowed" then it must exist. If you've "allowed" something that doesn't exist or no longer exists because the plugin is no longer installed then the service will fail to start.

> The logical question after that - let's restrict analyzers, normalizers, ... everything essentially cuz, well, this is a risk, right?

That's fair, except I would say it's not exactly about risk or trust, and instead about very fine grained configurability of the deployed system for an advanced user. My intuition is that we should enable configurability in most cases provided it doesn't complicate the default experience or add significant tech debt. I'm thinking this case passes that test because this setting can remain unspecified in most cases, and registering processors uniquely by name is already a concept in the system so putting a filter on that list seems like not a significant expansion an existing concept.

Reply:
> That's fair, except I would say it's not exactly about risk or trust, and instead about very fine grained configurability of the deployed system for an advanced user. 

I disagree with that: users are in control, they are aware of what they install and what system provides. We are hijacking that - we are fronting the extensibility (processors) by core (allow list), completely ignoring the reasons why we have plugins (and extensions!) in the first place. 

So, I would suggested to take a step back and address the problem at large:
 - we have many extensibility points (processors, aggregators, analyzers, normalizers, codecs (!), ...)
 - do we want to front that by allow list (in core) or not?

Although you are specifically targeting processors here, the scope is much larger. I would like for us to settle on the general approach how to deal with that.

Reply:
> we are fronting the extensibility (processors) by core (allow list)

We have extensibility points. Why have allow lists at all when you can just choose which things to install in the first place? That's a great point! The short answer is that today users are in fact not in control. For example, there is no supported way to not install a module. Bundled plugins (core or external repos) install extension points but the only knob a user has for a given plugin is install the entire thing or none of it. I think this is the problem at large. Do you agree?

There are two specific proposals to solve this problem discussed so far:

- Add one-off allow lists in core for each extension point as needed (this is the approach started by this PR)
- Make configurability about what to install the responsibility of each plugin/module suggested [here](https://github.com/opensearch-project/OpenSearch/pull/14479#discussion_r1649143484)

Reply:
>  Bundled plugins (core or external repos) install extension points but the only knob a user has for a given plugin is install the entire thing or none of it. I think this is the problem at large. Do you agree?

I agree about entire thing or none of it, but this is problem that plugins should solve, not us. Fe. because we have **no idea** of how plugins implement things, I could easily imagine when disallowing one processor could impact the ones which are allowed (delegation / pipelining /  ...)

Reply:
> Fe. because we have no idea of how plugins implement things

That's true, but individual processors are already top-level, user-facing concepts that end users can reference in isolation, so processor implementations must provide a reasonable experience if used by themselves.

I still think a core-level allow list for ingest processors is reasonable, mostly because using named processors is part of the core API. I think it is better than the alternative of implementing separate settings in each of `ingest-common`, `ingest-user-agent`, `ingest-geo-ip`, and `neural-search` in order to meet the requirement I listed above.

Could we end up building an allowlist setting for every extension point that registers named extensions? Maybe. I would say we can take it on a case-by-case basis. It probably won't make sense for everything.

Reply:
> That's true, but individual processors are already top-level, user-facing concepts that end users can reference in isolation, so processor implementations must provide a reasonable experience if used by themselves.

That is the thing, none of the processors is subject of undesigned exclusions / inclusions list, by supporting that where is actually is implemented would make perfect sense. 

> Could we end up building an allowlist setting for every extension point that registers named extensions? Maybe. I would say we can take it on a case-by-case basis. It probably won't make sense for everything.

Case by case won't apply here (that would be a mess), we are clearly having a very generic problem: there are external pieces coming to core that we try to impose the control on.

Reply:
I see https://github.com/opensearch-project/custom-codecs/issues/148 as slightly different because there is no concept of "experimental" codecs today, which is what that issue is trying to implement. We don't currently have a use case for defining experimental codecs directly in the core (to my knowledge), so I'm hesitant about adding that concept only to support a use case needed by custom-codecs. If the request was for a setting to select/restrict supported codecs, selecting from the union of plugin-provided and core-provided codecs, then I think I would be fine with the same pattern here.

> Case by case won't apply here (that would be a mess)

The requirement is to enable none of the processors in `ingest-common`, `ingest-user-agent`, and `ingest-geo-ip` and a subset of processors in `neural-search`. Practically speaking, isn't defining a distinct setting in each of those places more of a mess than a single core-defined setting?

Reply:
I see that the custom distribution requires a feature to turn off few processors and they are packaged as modules today in the distribution. 

My thoughts, I would like an answer to "does exposing a knob to choose/allow list/deny list processors help the community". Thinking about it I'm not convinced enough reading through the feature request.[1]

> by supporting that where is actually is implemented would make perfect sense.
I agree with @reta on this, the implementation lies far away from where its getting excluded. I would prefer to have the provider make the decision what to provide(processors in this case).

If we take defining plugin by plugin approach, I see we have to re-implement the setting as suggested by @andrross . 

I have another way, what if we define the setting  `ingest.processors.allowlist` and let the plugins filter out the processors they'd like to provide ? 
The upside is, `:server` module is not bloated with custom configuration and code.
The downside is, the plugins can choose to ignore the configuration, a.k.a no real enforcement of setting by OpenSearch. Is this a reasonable downside we can live with? 


[1] https://github.com/opensearch-project/OpenSearch/issues/14439

Reply:
> The downside is, the plugins can choose to ignore the configuration, a.k.a no real enforcement of setting by OpenSearch

@saratvemulapalli Eh, I think that's a pretty big downside. This would be a pretty obscure setting, and there is an extremely high likelihood that it would be unintentionally missed by plugin developers. I'd be more inclined to implement the setting in `neural-search` and then just `rm -rf` the unneeded modules.

Reply:
Yeah I agree, the reasonable approach is defining a setting per plugin, though its not ideal to define multiple settings for your use case, it helps the plugin dev to choose which related application business logic they'd like to disable. 

On the side note, can we move [ingest-user-agent](https://github.com/opensearch-project/OpenSearch/tree/main/modules/ingest-user-agent) to ingest-common?

Reply:
Let me move this allowlist implementation into `ingest-common` and separately work on a setting for neural-search.

> On the side note, can we move [ingest-user-agent](https://github.com/opensearch-project/OpenSearch/tree/main/modules/ingest-user-agent) to ingest-common?

@saratvemulapalli Great question! I think the answer is probably yes, but that is worth a separate issue for discussion.